src/expr.cpp

```diff
@@ -6177,61 +6185,68 @@ int ConstExpr::EstimateCost() const { return 0; }
 
 ConstExpr *ConstExpr::Instantiate(TemplateInstantiation &templInst) const { return new ConstExpr(this, pos); }
 
-void ConstExpr::Print(Indent &indent) const {
-    indent.Print("ConstExpr", pos);
-
-    printf("[%s] (", GetType()->GetString().c_str());
+std::string ConstExpr::GetValuesAsStr(const std::string &separator) const {
```

Comment:
What do you think about changing the method name to `GetString` and making a default value for argument? It makes sense for debugging (`Print` interface is annoying to use) and matches method name of `Type` classes hierarchy.

Reply:
> `Print` interface is annoying to use

It's here for you:
https://github.com/ispc/ispc/blob/c72af29f0bd82678918b941654aff892fbd7dc8a/src/ast.h#L175-L176

Reply:
And here how it doesn't work for interactive debugging:
```bash
$ cat max.ispc
int foo(int a, int b) { return max(a, b); }
$ gdb --args ./bin/ispc --target=host max.ispc
(gdb) b ReturnStmt::EmitCode
Breakpoint 1 at 0x3064424: file /localdisk2/anurmukh/applications.compilers.ispc.core/src/stmt.cpp, line 2957.
(gdb) r                                                                                                                                                                                                                                                     Starting program: /localdisk2/anurmukh/applications.compilers.ispc.core/build-debug/bin/ispc --target=host max.ispc
Breakpoint 1, ispc::ReturnStmt::EmitCode (this=0x55555ae9ed50, ctx=0x7fffffffd440) at /localdisk2/anurmukh/applications.compilers.ispc.core/src/stmt.cpp:2957
2957        if (!ctx->GetCurrentBasicBlock())
(gdb) p expr
$1 = (ispc::Expr *) 0x55555ae9ecf0
(gdb) call expr->Print()
Cannot evaluate function -- may be inlined
```
ISPC is configured with `cmake ../ -DCMAKE_BUILD_TYPE=Debug`.

Moreover, there is an interesting observation related to `Print` functionality here https://github.com/ispc/ispc/pull/2716#issuecomment-1850921990 that it can interfere with compiler logic. Of course, this comment is completely unrelated to this PR.

Reply:
Interesting, I implemented this function to use in the debugger and it did work for me. Worth investigating why it's not working.

Reply:
It works for me on macOS with `lldb`:
```
DMITRYs-MacBook-Pro-2:~/ispc/dbabokin_ispc/build (main)> lldb -- bin/ispc --target=host max.ispc -o max.o
(lldb) target create "bin/ispc"
Current executable set to '/Users/dbabokin/ispc/dbabokin_ispc/build/bin/ispc' (arm64).
(lldb) settings set -- target.run-args  "--target=host" "max.ispc" "-o" "max.o"
(lldb) b ReturnStmt::EmitCode
Breakpoint 1: where = ispc`ispc::ReturnStmt::EmitCode(ispc::FunctionEmitContext*) const + 56 at stmt.cpp:2957:10, address = 0x000000010024e354
(lldb) r
Process 71055 launched: '/Users/dbabokin/ispc/dbabokin_ispc/build/bin/ispc' (arm64)
Process 71055 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x000000010024e354 ispc`ispc::ReturnStmt::EmitCode(this=0x0000600002af7cf0, ctx=0x000000016fdfda88) const at stmt.cpp:2957:10
   2954	ReturnStmt::ReturnStmt(Expr *e, SourcePos p) : Stmt(p, ReturnStmtID), expr(e) {}
   2955
   2956	void ReturnStmt::EmitCode(FunctionEmitContext *ctx) const {
-> 2957	    if (!ctx->GetCurrentBasicBlock())
   2958	        return;
   2959
   2960	    if (ctx->InForeachLoop()) {
Target 0: (ispc) stopped.
(lldb) p expr->Print()
`-FunctionCallExpr <stdlib.ispc:81.86 - 81.113> [varying float16]
  |-(func) FunctionSymbolExpr <stdlib.ispc:81.86 - 81.110> [ varying float16(varying unsigned int16 )] function name: __halfbits_varying_int16
  `-(args) ExprList <stdlib.ispc:81.111 - 81.112>
    `-SymbolExpr <stdlib.ispc:81.111 - 81.112> [varying unsigned int16] symbol name: a
```

Reply:
@nurmukhametov does adding `__attribute__((used))` to `Print()` declaration help in your environment?

Reply:
For me it doesn't help.
```
Breakpoint 1, ispc::BinaryExpr::GetValue (this=0x55555b968d20, ctx=0x7fffffffd238)
    at /export/users/aneshlya/ispc/applications.compilers.ispc.core/src/expr.cpp:2101
2101        if (!arg0 || !arg1) {
(gdb) p Print()
Too few arguments in function call.
```

Reply:
Actually, no need for this attribute. But we don't need security hardening linker options in the debug build. Specifically this one is doing the harm:
https://github.com/ispc/ispc/blob/c72af29f0bd82678918b941654aff892fbd7dc8a/CMakeLists.txt#L615
I verified it on Linux.

Reply:
Fixed in https://github.com/ispc/ispc/pull/2797